testing/unittests/jlibtests.cpp

```diff
@@ -3611,4 +3614,144 @@ CPPUNIT_TEST_SUITE_REGISTRATION( JLibUnicodeTest );
 CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( JLibUnicodeTest, "JLibUnicodeTest" );
 
 
+class JLibSecretsTest : public CppUnit::TestFixture
+{
+public:
+    CPPUNIT_TEST_SUITE(JLibSecretsTest);
+        CPPUNIT_TEST(setup);
+        CPPUNIT_TEST(testUpdate);
+    CPPUNIT_TEST_SUITE_END();
+
+    //Each test creates a different instance of the class(!) so member values cannot be used to pass items
+    //from one test to another
+    StringBuffer secretRoot;
+
+protected:
+    void checkSecret(const char * secret, const char * key, const char * expectedValue)
+    {
+        Owned<const IPropertyTree> match = getSecret("testing", secret);
+        CPPUNIT_ASSERT(match);
+        const char * secretValue = match->queryProp(key);
+        CPPUNIT_ASSERT(secretValue);
+        CPPUNIT_ASSERT_EQUAL_STR(secretValue, expectedValue);
+    }
+
+    bool hasSecret(const char * name)
+    {
+        Owned<const IPropertyTree> match = getSecret("testing", name);
+        return match != nullptr;
+    }
+
+    void initPath()
+    {
+        char cwd[1024];
+        CPPUNIT_ASSERT(GetCurrentDirectory(1024, cwd));
+        secretRoot.set(cwd).append(PATHSEPCHAR).append("unittest-secrets");
+        secretRoot.append(PATHSEPCHAR).append("testing"); // catgegory
+    }
+
+    void setup()
+    {
+        char cwd[1024];
+        CPPUNIT_ASSERT(GetCurrentDirectory(1024, cwd));
+        secretRoot.append(cwd).append(PATHSEPCHAR).append("unittest-secrets");
+
+        recursiveRemoveDirectory(secretRoot);
+        CPPUNIT_ASSERT(recursiveCreateDirectory(secretRoot.str()));
+        setSecretMount(secretRoot);
+        setSecretTimeout(100); // Set the timeout so we can check it is working.
+
+        secretRoot.append(PATHSEPCHAR).append("testing"); // catgegory
+        CPPUNIT_ASSERT(recursiveCreateDirectory(secretRoot.str()));
+    }
+
+    void testUpdate()
+    {
+        initPath(); // secretRoot needs to be called for each test
+
+        CPPUNIT_ASSERT(!hasSecret("secret1"));
+        writeTestingSecret("secret1", "value", "secret1Value");
+        //Secret should not appear yet - null should be cached.
+        CPPUNIT_ASSERT(!hasSecret("secret1"));
+
+        Owned<ISyncedPropertyTree> secret2 = resolveSecret("testing", "secret2", nullptr, nullptr);
+        CPPUNIT_ASSERT(!secret2->isValid());
+        CPPUNIT_ASSERT(!secret2->isStale());
+
+        MilliSleep(50);
+        //Secret should not appear yet - null should be cached.
+        CPPUNIT_ASSERT(!hasSecret("secret1"));
+        CPPUNIT_ASSERT(!secret2->isValid());
+        CPPUNIT_ASSERT(!secret2->isStale());
+
+        MilliSleep(100);
+        //Secret should not be updated - enough time has passed
```

Comment:
Perhaps "Secret has not automatically been updated - but enough time has passed"?